pkg/apis/pipeline/v1/matrix_types.go

```diff
@@ -56,9 +56,79 @@ type MatrixInclude struct {
 // Combination is a map, mainly defined to hold a single combination from a Matrix with key as param.Name and value as param.Value
 type Combination map[string]string
 
-// Combinations is a list of combinations
+// Combinations is a Combination list
 type Combinations []Combination
 
+// FanOut generates Combinations, which is a list of mapped Combination by fanning out matrix
+// Parameters including Matrix Include Parameters
+func (m *Matrix) FanOut() Combinations {
+	var combinations, includeCombinations Combinations
+
+	// Generate include combinations from Matrix Include Parameters
+	includeCombinations = m.fanOutIncludeParams()
+
+	if m.hasInclude() && !m.hasParams() {
+		// If there are only Matrix Include Parameters return explicit combinations
+		return includeCombinations
+	}
+
+	// Generate combinations from Matrix Parameters
+	for _, parameter := range m.Params {
+		combinations = combinations.fanOutMatrixParams(parameter)
+	}
+
+	// Override the current combination with the include combination if any of the include parameters are
+	// found in combinations with matching parameter name and values
+	for _, paramCombination := range combinations {
+		for _, includeCombination := range includeCombinations {
+			if paramCombination.inCombination(includeCombination) {
+				for name, val := range includeCombination {
+					paramCombination[name] = val
+				}
+			}
+		}
+	}
+
+	// Generate a new combination for the missing include parameter value
+	for _, includeCombination := range includeCombinations {
+		if combinations.isIncludeNewCombination(includeCombination) {
+			combinations = append(combinations, includeCombination)
+		}
+	}
+
+	return combinations
+}
+
+// inCombination returns true if the include parameter name and value exists in combinations
+func (c Combination) inCombination(includeCombination Combination) bool {
+	for name, val := range includeCombination {
+		if _, exist := c[name]; exist {
+			if c[name] != val {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// isIncludeNewCombination returns true if the include parameter name exists but the value is
+// missing from combinations
+func (cs Combinations) isIncludeNewCombination(includeCombination map[string]string) bool {
```

Comment:
```suggestion
// addNewCombination returns true if the include parameter name exists but the value is
// missing from combinations: in this case, we need to add a new combination
func (cs Combinations) addNewCombination(includeCombination map[string]string) bool {
```

Reply:
Renamed to `shouldAddNewCombination` to emphasize it's returning a bool.
`addNewCombination` makes me think the function is adding a combination.